### PR TITLE
Add hex_tarball

### PR DESCRIPTION
**Pull request by wojtekmach**
TODO:

- [x] Approve PR for hex (hexpm/hex#518)
- [x] Approve PR for hexpm (hexpm/hexpm#641)

### CODE REVIEW

File: src/hex_tarball.erl [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+                            {<<"Makefile">>, <<"make">>},
+                            {<<"Makefile.win">>, <<"make">>}
+                           ]).
+-include_lib("kernel/include/file.hrl").
+
+-type checksum() :: binary().
+-type contents() :: #{filename() => binary()}.
+-type filename() :: string().
+-type files() :: [filename()] | contents().
+-type metadata() :: map().
```
Comment (wojtekmach) [inline src/hex_tarball.erl]: currently metadata has binary keys, eventually I'd like to have atom keys but not in this PR as it requires more changes in hexpm and has security implications.

File: src/hex_tarball.erl [hunk 2]
```diff
@@ -0,0 +1,455 @@
+-module(hex_tarball).
```
Comment (wojtekmach) [inline src/hex_tarball.erl]: some other ideas for name are:

- `hex`
- `hex_package`
- `hex_erl_{tarball|package}`

I kinda like `hex_package` the most because then `hex_package:tarball()` type looks better than `hex_tarball:tarball()` :-)

other ideas?

PS I briefly considered having type `t() :: #{checksum := ..., contentes := ..., metadata := ...}` which would be really nice because it exactly matches the spec, but this type would have been used only when unpacking to memory (when unpacking to disk we don't return contents)

PS2 I wonder if there's gonna be `hex_erl` module at all. I think we could add later `hex_erl:version/1` so that when we vendor this it could be used to generate stuff like:

```
cat vendored_hex_tarball.erl
%% Vendored from hex_erl v0.1.0-dev, do not edit manually
```

File: src/hex_tarball.erl [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+-export([create/2, unpack/2, format_checksum/1, format_error/1, gzip/1, normalize_requirements/1]).
```
Comment (wojtekmach) [inline src/hex_tarball.erl]: gzip/1 and normalize_requirements/1 are only exposed for testing (gzip is currently tested in Hex, I'll move property test over in a separate PR). What's the best way to expose functions just for tests and not in prooduction?

File: test/hex_tarball_tests.erl [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    ok = rm_rf(TmpDir),
+    ok = file:make_dir(TmpDir),
+    Dir = TmpDir ++ "/" ++ integer_to_list(erlang:unique_integer()),
+    ok = file:make_dir(Dir),
+    file:set_cwd(Dir),
+    apply(F, []),
+    file:set_cwd(Old).
+
+rm_rf(Path) ->
+    [] = os:cmd("rm -rf " ++ Path),
```
Comment (wojtekmach) [inline test/hex_tarball_tests.erl]: this is ugly, need to find a cleaner way.

File: src/hex_tarball.erl [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+-define(VERSION, <<"3">>).
+-define(TARBALL_MAX_SIZE, 8 * 1024 * 1024).
```
Comment (wojtekmach) [inline src/hex_tarball.erl]: in a separate PR we'll make tarball max size configurable via options to create and unpack

File: src/hex_tarball.erl [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+
+        {ok, Contents} ->
+            {ok, #{checksum => Checksum, metadata => Metadata, contents => Contents}};
+
+        {error, Reason} ->
+            {error, {inner_tarball, Reason}}
+    end.
+
+copy_metadata_config(Output, MetadataBinary) ->
+    ok = file:write_file(Output ++ "/hex_metadata.config", MetadataBinary).
```
Comment (wojtekmach) [inline src/hex_tarball.erl]: this is somewhat Hex specific feature, not sure if it makes sense for other build tools.

File: test/hex_tarball_tests.erl (hunk 4 above)
Comment (tsloughter) [reply test/hex_tarball_tests.erl]: Less important since it is a test, but in case anyone wants to contribute from Windows it'll need support. So something like https://github.com/erlang/rebar3/blob/master/src/rebar_file_utils.erl#L170

File: src/hex_tarball.erl (hunk 3 above)
Comment (tsloughter) [reply src/hex_tarball.erl]: Wrap it in an `ifdef`. rebar3 sets `TEST` when running tests so you can do:

```erlang
-ifdef(TEST).
-export([gzip]).
-endif.
```

**Comment by wojtekmach** [comment]
@ericmj a while back we talked about writing a tool that runs over all published packages to make sure we have not introduced a regression. Here's a stab at it: https://github.com/wojtekmach/hex_test/blob/master/lib/hex_test.ex#L1. With a few caveats due to broken packages (documented in that file) it looks like `hex_tarball` is compatible with `Hex.Tar` 👍

File: src/hex_tarball.erl (hunk 2 above)
Comment (ericmj) [reply src/hex_tarball.erl]: I think I prefer the current name since the module is specifically about the package tarball and we can have other modules like the package registry etc.. I was thinking `hex_erl` for the project/application name and use the `hex_` namespace.

File: src/hex_tarball.erl (hunk 6 above)
Comment (ericmj) [reply src/hex_tarball.erl]: I think we can keep it for now.

File: .travis.yml [hunk 7]
```diff
@@ -2,3 +2,6 @@ language: erlang
 sudo: false
 otp_release:
   - 20.2
```
Comment (ericmj) [inline .travis.yml]: We need more OTP versions here.

@tsloughter What is rebar3's minimum supported OTP versions?

File: src/hex_tarball.erl [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+%% |...compressed blocks...| (more-->)
+%% +=======================+
+%%
+%% +---+---+---+---+---+---+---+---+
+%% |     CRC32     |     ISIZE     |
+%% +---+---+---+---+---+---+---+---+
+gzip(Uncompressed) ->
+    Compressed = gzip_no_header(Uncompressed),
+    Header = <<31, 139, 8, 0, 0, 0, 0, 0, 0, 0>>,
+    Crc = erlang:crc32(Uncompressed),
```
Comment (ericmj) [inline src/hex_tarball.erl]: Just `Crc = crc32(Uncompressed)` works, right?

File: test/prop_hex_tarball.erl [hunk 9]
```diff
@@ -0,0 +1,16 @@
+-module(prop_hex_tarball).
+-include_lib("proper/include/proper.hrl").
```
Comment (ericmj) [inline test/prop_hex_tarball.erl]: 👍

**Comment by ericmj** [comment]
I will take a look at the broken packages and figure out what to do with them.

File: rebar.config [hunk 10]
```diff
@@ -1,2 +1,8 @@
 {erl_opts, [debug_info]}.
-{deps, []}.
\ No newline at end of file
+{project_plugins, [rebar3_proper]}.
+{deps, []}.
+{profiles,
+    [{test, [
+        {deps, [{proper, "1.2.0"}, recon]}
```
[the remark this answers is not in the document]
Comment (wojtekmach) [reply rebar.config]: Good catch, will remove

File: src/hex_tarball.erl (hunk 8 above)
Comment (wojtekmach) [reply src/hex_tarball.erl]: no, most functions from `erlang` are auto-imported but not all and not this one (TIL)

File: .travis.yml (hunk 7 above)
Comment (wojtekmach) [reply .travis.yml]: @ericmj how about mirroring Hex's travis, so 20.2, 19.3, 18.3, and maybe 17.5?

Looks like proper plugin doesn't work on 17.5 (https://travis-ci.org/hexpm/hex_erl/jobs/348937049), I'll investigate. If it's gonna be tricky I think it's OK to skip 17.5 here - hex_erl will still be tested on 17.5 anytime we re-vendor it on Hex (e.g. before each hex_erl release)

File: src/hex_tarball.erl (hunk 2 above)
Comment (wojtekmach) [reply src/hex_tarball.erl]: 👍

File: rebar.lock [hunk 11]
```diff
@@ -0,0 +1 @@
+[].
```
[the remark this answers is not in the document]
Comment (ferd) [reply rebar.lock]: they are not. We only lock the default stuff.

File: src/hex_filename.erl [hunk 12]
```diff
@@ -0,0 +1,55 @@
+% Excerpt from https://github.com/erlang/otp/blob/OTP-20.0.1/lib/stdlib/src/filename.erl#L761-L788
+% with modifications for changing local function calls to remote function calls
+% to the `filename` module, for the functions `pathtype/1`, `split/1`, and `join/1`
```
Comment (ferd) [inline src/hex_filename.erl]: Why is that required?

File: src/hex_tarball.erl [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+
+        {error, {_Line, safe_erl_term, Reason}, _Line2} ->
+            {error, {metadata, Reason}}
+    end.
+
+safe_to_string(Binary) ->
+    case unicode:characters_to_list(Binary) of
+        List when is_list(List) -> List;
+        {error, _, _} -> binary_to_list(Binary);
+        {incomplete, _, _} -> binary_to_list(Binary)
```
Comment (ferd) [inline src/hex_tarball.erl]: these are both potentially unsafe as well; i.e. what you may get is an invalid byte sequence out of it which is why things don't convert, and by using binary_to_list you're just converting these invalid byte sequences into unrelated codepoint values.

Unless you have specific cases where this happens, it would be safer to bail out if the conversion cannot work.

File: src/hex_tarball.erl [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+
+maybe_update_with(Key, Fun, Map) ->
+    case maps:is_key(Key, Map) of
+        true ->
+            Value = maps:get(Key, Map),
+            maps:put(Key, apply(Fun, [Value]), Map);
+
+        false ->
+            Map
+    end.
```
Comment (ferd) [inline src/hex_tarball.erl]: would probably more efficient with

```erlang
case maps:find(Key, Map) of
     {ok, Value} -> maps:put(Key, Fun(Value), Map);
     error -> Map
end
%% OR
case Map of
    #{Key := Value} -> Map#{Key := Fun(Value)};
    _ -> Map
end. 
```
- skip one lookup round
- literal function application is a bit faster than `apply/2` as well given we know the argument count

File: test/prop_hex_tarball.erl [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+
+prop_symmetric() ->
+    ?FORALL(Binary, binary(),
```
Comment (ferd) [inline test/prop_hex_tarball.erl]: you should probably `collect(byte_size(Binary), ...)` here and see what you get. I think chances are you'll mostly get short binaries and the test could be more trustworthy by toying with the generator size to get something that may reach a few MBs here and there.

File: src/hex_filename.erl (hunk 12 above)
Comment (ericmj) [reply src/hex_filename.erl]: Because we only vendor one function instead of the whole file.

Comment (ferd) [reply src/hex_filename.erl]: I thought you could just use the stdlib functions without vendoring? What prevents that?

File: src/hex_tarball.erl (hunk 14 above)
Comment (wojtekmach) [reply src/hex_tarball.erl]: yup, thank you! I went with option 1 because `#{Key := Value}` is not available on OTP 17

File: src/hex_tarball.erl (hunk 13 above)
Comment (wojtekmach) [reply src/hex_tarball.erl]: you're right, they're unsafe. I re-run new tarball code on all tarballs and the good news is none hit `incomplete` clause but 206 (out of ~30k) hit `error` clause.

the first one to hit it is blacksmith-0.1.0:

`
no case clause matching: {:error, '{<<"app">>,<<"blacksmith">>}.\n{<<"build_tools">>,[<<"mix">>]}.\n{<<"contributors">>,[<<"Bruce Tate">>,<<"Eric Meadows-J', <<246, 110, 115, 115, 111, 110, 34, 62, 62, 93, 125, 46, 10, 123, 60, 60, 34, 100, 101, 115, 99, 114, 105, 112, 116, 105, 111, 110, 34, 62, 62, 44, 10, 32, 60, 60, 34, 69, 108, 105, 120, 105, 114, 32, 102, 97, 107, ...>>}
`

:-)

here's whole list: https://gist.github.com/wojtekmach/cf5287ace7d464f801d86b962eed71c0

since these code is for unpacking and affects existing packages I don't think we should crash. I suppose we could warn or something, but since it's < 1% and looks like pretty old packages I think it's ok to leave as is.

File: test/prop_hex_tarball.erl (hunk 15 above)
Comment (wojtekmach) [reply test/prop_hex_tarball.erl]: good catch. I got at most 40 bytes on a recent run. Will investigate and improve later.

thanks for this and your other suggestions!

File: src/hex_filename.erl (hunk 12 above)
Comment (wojtekmach) [reply src/hex_filename.erl]: I've added a rationale to that file which is that `safe_relative_path` was not present in previous OTP releases.

File: src/hex_tarball.erl (hunk 13 above)
Comment (ferd) [reply src/hex_tarball.erl]: ah those were probably latin1. What do you get if you try `unicode:characters_to_list(Str, latin1, unicode)` ?

Comment (wojtekmach) [reply src/hex_tarball.erl]: good call:

```erlang
1> unicode:characters_to_list(<<69, 114, 105, 99, 32, 77, 101, 97, 100, 111, 119, 115, 45, 74, 246, 110, 115, 115, 111, 110>>).
{error,"Eric Meadows-J",<<"önsson">>}
2> unicode:characters_to_list(<<69, 114, 105, 99, 32, 77, 101, 97, 100, 111, 119, 115, 45, 74, 246, 110, 115, 115, 111, 110>>, latin1).
"Eric Meadows-Jönsson"
```

I'm gonna add fallback to latin1 👍

File: test/prop_hex_tarball.erl (hunk 15 above)
Comment (ferd) [reply test/prop_hex_tarball.erl]: you can try doing something like `?SIZED(Size, resize(Size*100, binary()))` to get 100x longer values based on the current generator size, so you can still get small binaries, but larger ones as well. Play with such values to get broader generation.

File: src/hex_tarball.erl (hunk 13 above)
Comment (wojtekmach) [reply src/hex_tarball.erl]: @ferd well that was fun.

I've made the conversion more explicit in d08c2d51487d6f90645fcc31dada06ae62952952, that's a nice improvement even though in that particular case it has the same outcome as just calling `binary_to_list(Binary)` for that particular latin1-encoded binary.

To properly fix up encoding we'd need to do:

```erlang
unicode:characters_to_list(
  unicode:characters_to_binary(Binary, latin1, unicode),
  latin1
)
```

but then it still doesn't quite work, this test would fail: https://github.com/hexpm/hex_erl/blob/43b8b3fb650da5bede609ecfbea6a87812d27f5c/test/hex_tarball_tests.erl#L120 because we'd be missing the `/utf8` sigil in the binary literal in the result.

I'm giving up on this for now but I'll likely revisit this later.
